### DISCUSSION TITLE
How can I close and reopen the database? (delete data issue)

### DISCUSSION THREAD

**Discussion by lifeapps-42** [simolus3/drift#2513] [Q&A] [answered]
I'm facing an issue, that I beleve could be resolved with database reoppening.

We have several scenarios when need to clean all data from local DB. Now I just clean all rows with this:

```dart
  Future<void> wipeAllData() async {
    await customStatement('PRAGMA foreign_keys = OFF');
    try {
      await transaction(() async {
        for (final table in allTables) {
          await delete(table).go();
        }
      });
    } on Exception catch (e, st) {
      ref.read(dbLoger).log(e.toString(), st: st);
    } finally {
      await customStatement('PRAGMA foreign_keys = ON');
      ref.read(dbLoger).log('Database deleted');
    }
  }
```

But sometimes right after wiping out, db refuses to store any new data with `ForeignKeyConstrainsException`

Reloading app helps, an it resumes working just fine.

So, my idea is to close and then reopen database to resolve this issue.

Closing is pretty straightforvard: `db.close()`, but how to reopen? Do i need to create a new Executor instance?

**Comment by simolus3** [comment]
Databases currently can't be re-opened. I've considered adding that functionality after first seeing your issue, but I think it's probably best if "closed" is a terminal state.

We should probably figure out the issue of the foreign key exception instead. Can you run [`pragma foreign_key_check`](https://www.sqlite.org/pragma.html#pragma_foreign_key_check) after re-enabling them and see if the error happens there already?

**Reply by lifeapps-42** [reply]
Thank you for your response!

Do you want me to modify my function this way? 
```dart
 Future<void> wipeAllData() async {
    await customStatement('PRAGMA foreign_keys = OFF');
    try {
      await transaction(() async {
        for (final table in allTables) {
          await delete(table).go();
        }
      });
    } on Exception catch (e, st) {
      ref.read(dbLoger).log(e.toString(), st: st);
    } finally {
      await customStatement('PRAGMA foreign_keys = ON');
      //add this line?
      await customStatement('PRAGMA foreign_key_check');
      ref.read(dbLoger).log('Database deleted');
    }
  }
```

**Reply by simolus3** [reply]
Basically yes, but it would be helpful to print the results as well:

```dart
  await customStatement('PRAGMA foreign_keys = ON');
  // add this to the finally block:
  final rows = await customSelect('pragma foreign_key_check;').get();
  for (final row in rows) {
    print('Foreign key issue: ${row.data}');
  }
```

Note that this won't fix the problem, but it might help point out what to look at next. Perhaps that also needs to go before the `PRAGMA foreign_keys = ON` line, I'm not sure.

**Reply by lifeapps-42** [reply]
Got it. I'll try this. Thx)

**Comment by lifeapps-42** [comment, marked as answer]
After some futher investigation I found out an issue on my side. Basicly, I was facing tricky race condition)

Thank you for your help! We may close this thread now)

**Comment by darkomenx** [comment]
Is there a new way to solve to issue with keys constraints ?

I need to reset a database at the runtime (then disable foreign_keys -> delete data -> enable foreign keys -> insert data) which as started with foreign keys activate
